chainer/functions/math/sparse_matmul.py

```diff
@@ -0,0 +1,429 @@
+            raise ValueError('ndim of sp_row and sp_col must be one or two.')
+        for i in range(sp_row.ndim):
+            if sp_row.shape[i] != sp_col.shape[i]:
+                _msg = 'shape of sp_row and sp_col must be the same.'
+                raise ValueError(_msg)
+        if len(sp_shape) != 2:
+            raise ValueError('len(sp_shape) must be two.')
+        self.sp_row = sp_row  # ((nb,) ldnz)
+        self.sp_col = sp_col  # ((nb,) ldnz)
+        self.sp_shape = sp_shape  # (_m, _n) when transc is False
+        self.transa = transa
+        self.transb = transb
+        self.transc = transc
+        self.dtype = dtype
+
+    def check_type_forward(self, in_types):
+        type_check.expect(in_types.size() == 2)
+        a_type, b_type = in_types
+        # a_type.shape: ((nb,) _m, _k) when transa is False
+        # b_type.shape: ((nb,) _k, _n) when transb is False
+        a_m_axis, a_k_axis = -2, -1
+        b_k_axis, b_n_axis = -2, -1
+        sp_m_axis, sp_n_axis = -2, -1
+        if self.transa:
+            a_m_axis, a_k_axis = -1, -2
+        if self.transb:
+            b_k_axis, b_n_axis = -1, -2
+        if self.transc:
+            sp_m_axis, sp_n_axis = -1, -2
+        type_check.expect(
+            a_type.dtype.kind == 'f',
+            b_type.dtype.kind == 'f',
+            a_type.ndim >= 2,
+            a_type.ndim <= 3,
+            a_type.ndim == b_type.ndim,
+            a_type.shape[a_m_axis] == self.sp_shape[sp_m_axis],
+            b_type.shape[b_n_axis] == self.sp_shape[sp_n_axis],
+            a_type.shape[a_k_axis] == b_type.shape[b_k_axis],
+        )
+        a_ndim = type_check.eval(a_type.ndim)
+        if a_ndim == 3:
+            type_check.expect(
+                a_type.shape[0] == self.sp_row.shape[0],
+                b_type.shape[0] == self.sp_row.shape[0],
+            )
+
+    def forward(self, inputs):
+        self.retain_inputs((0, 1))
+        a, b = inputs
+        c = _sparse_matmul_gradsp(a, b,
+                                  self.sp_row, self.sp_col, self.sp_shape,
+                                  self.transa, self.transb, self.transc,
+                                  self.dtype)
+        return utils.force_array(c),
+
+    def backward(self, indexes, grad_outputs):
+        a, b = self.get_retained_inputs()
+        g_sp, = grad_outputs
+
+        g_a = None
+        if 0 in indexes:
+            g_a = SparseMatMul(self.sp_row, self.sp_col, self.sp_shape,
+                               self.transc, not self.transb, self.transa,
+                               dtype=a.dtype).apply((g_sp, b))[0]
+
+        g_b = None
+        if 0 in indexes:
+            g_b = SparseMatMul(self.sp_row, self.sp_col, self.sp_shape,
+                               not self.transc, self.transa, not self.transb,
+                               dtype=b.dtype).apply((g_sp, a))[0]
+
+        return g_a, g_b
+
+
+def sparse_matmul(a, b, transa=False, transb=False):
```

Comment:
I wonder if we should rename this function `batch_sparse_matmul` for consistency with https://docs.chainer.org/en/stable/reference/generated/chainer.functions.batch_matmul.html, or `batch_sparse_dense_matmul` to be more explicit, what do you think?

Reply:
`batch_matmul` is dperecated as written in the document.

```
Deprecated since version v3.0.0:
batch_matmul is deprecated. Use matmul instead.
```

Original numpy `matmul` already acts like "batch_matmul" when input is 3-dim. Therefore, I think current terminology is already consistent with `matmul` API.

Reply:
Ah, thanks a lot for commenting on that. Then I think we should definitely keep it as is.

Reply:
Thanks, @corochann , for your comment. I also think `sparse_matmul` is better as a name of this function.